### PR TITLE
[MM-20233] Adding cloud blackbox target discovery microservice.

### PR DESCRIPTION
**Pull request by stylianosrigas**
#### Summary
- Adding cloud blackbox target discovery microservice.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-30233

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -0,0 +1,325 @@
+package main
+
+import (
```
Comment (cpanato) [inline main.go:3]: would write 
```
import (
	"context"
	"fmt"
	"io/ioutil"
	"os"
	"path/filepath"
	"strings"

	log "github.com/sirupsen/logrus"
	"gopkg.in/yaml.v2"
	"github.com/aws/aws-sdk-go/aws"
	"github.com/aws/aws-sdk-go/aws/session"
	"github.com/aws/aws-sdk-go/service/route53"
	"github.com/pingcap/errors"

	corev1 "k8s.io/api/core/v1"
	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/client-go/kubernetes"
	"k8s.io/client-go/rest"
	"k8s.io/client-go/tools/clientcmd"
)
```

File: main.go [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+
+	"gopkg.in/yaml.v2"
+	corev1 "k8s.io/api/core/v1"
+	k8sErrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/pingcap/errors"
```
Comment (cpanato) [inline main.go]: are you sure this is the package you want?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+				blackBoxTargets = append(blackBoxTargets, fmt.Sprintf("%s:9090", *record.Name))
+			}
+		}
+	}
+
+	for _, target := range additionalTargets {
+		log.Infof("Adding additional target %s", target)
+		blackBoxTargets = append(blackBoxTargets, target)
+	}
+	log.Info("Returning Blackbox targets")
```
Comment (cpanato) [inline main.go:298]: nit: new line here

File: main.go [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+// isExcludedTarget checks if a Route53 record is in the excluded targets
+func isExcludedTarget(excludedTargets []string, record string) bool {
+	if len(excludedTargets) > 0 {
+		for _, target := range excludedTargets {
+			if target == record {
+				return true
+			}
+		}
+	}
+	return false
```
Comment (cpanato) [inline main.go:313]: nit: new line here

File: main.go [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return clientset, nil
```
Comment (cpanato) [inline main.go:234]: nit: new line

File: main.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+			return nil, err
+		}
+		return clientset, nil
+	}
+
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	clientset, err := kubernetes.NewForConfig(config)
```
Comment (cpanato) [inline main.go:229]: nit: new line

File: main.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+
+func main() {
+	envVars, err := validateAndGetEnvVars()
+	if err != nil {
+		log.WithError(err).Error("Environment variable validation failed")
+		err = sendMattermostErrorNotification(err, "Environment variable validation failed")
+		if err != nil {
+			log.WithError(err).Error("Failed to send Mattermost error notification")
+		}
+		return
```
Comment (cpanato) [inline main.go]: even if there is an error we don't care?

maybe we use the os.exit(0) when success and os.Exist(1) or another number when fail?

File: notification.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	req.Header.Set("X-Custom-Header", "aws-sns")
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrap(err, "failed tο send HTTP request")
+	}
+	defer resp.Body.Close()
+	return nil
```
Comment (cpanato) [inline notification.go:27]: nit: new line

File: notification.go [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+	payload := model.CommandResponse{
+		Username:    "Blackbox Target Discovery",
+		IconURL:     "https://upload.wikimedia.org/wikipedia/commons/thumb/3/38/Prometheus_software_logo.svg/1200px-Prometheus_software_logo.svg.png",
+		Attachments: []*model.SlackAttachment{attachment},
+	}
+	err := send(os.Getenv("MATTERMOST_ALERTS_HOOK"), payload)
+	if err != nil {
+		return errors.Wrap(err, "failed tο send Mattermost error payload")
+	}
+	return nil
```
Comment (cpanato) [inline notification.go:49]: nit: new line

File: main.go [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+	}
+
+	log.Info("Successfully updated Blackbox targets")
+	return nil
+}
+
+// getClientSet gets the k8s clientset
+func getClientSet(envVars *environmentVariables) (*kubernetes.Clientset, error) {
+	if envVars.DevMode == "true" {
+
```
Comment (cpanato) [inline main.go]: remove this blank line

File: main.go [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+		config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			return nil, err
+		}
+		clientset, err := kubernetes.NewForConfig(config)
+		if err != nil {
+			return nil, err
+		}
+		return clientset, nil
```
Comment (cpanato) [inline main.go:221]: nit: new line

File: main.go [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+
+		kubeconfig := filepath.Join(
+			os.Getenv("HOME"), ".kube", "config",
+		)
+
+		config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+		if err != nil {
+			return nil, err
+		}
+		clientset, err := kubernetes.NewForConfig(config)
```
Comment (cpanato) [inline main.go:216]: nit: new line

**Review by cpanato — CHANGES_REQUESTED** [review]
some nits and comments
thanks for this!

File: main.go (hunk 2 above)
Comment (stylianosrigas) [reply main.go]: Nice catch, thanks :)

**Comment by stylianosrigas** [comment]
> some nits and comments
> thanks for this!

Thanks for the review @cpanato. All fixed :)

File: main.go [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+	}
+
+	log.Info("Creating/updating Blackbox targets Prometheus secret")
+	_, err = createOrUpdateSecret(envVars.PrometheusNamespace, envVars.PrometheusSecretName, secret, clientset)
+	if err != nil {
+		return errors.Wrap(err, "failed to create the Blackbox targets Prometheus secret")
+	}
+
+	log.Info("Successfully updated Blackbox targets")
+
```
Comment (cpanato) [inline main.go:200]: can remove this extra line

**Review by cpanato — APPROVED** [review]

**Review by jwilander — APPROVED** [review]
LGTM

**Pull request merged**